src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceParameterService.cs

```diff
@@ -0,0 +1,688 @@
+        protected static async Task<ImmutableDictionary<Document, List<TInvocationExpressionSyntax>>> FindCallSitesAsync(
+            SemanticDocument document, IMethodSymbol methodSymbol, CancellationToken cancellationToken)
+        {
+            var methodCallSites = new Dictionary<Document, List<TInvocationExpressionSyntax>>();
+            var progress = new StreamingProgressCollector();
+
+            await SymbolFinder.FindReferencesAsync(
+                methodSymbol, document.Document.Project.Solution, progress: progress,
+                documents: null, FindReferencesSearchOptions.Default, cancellationToken).ConfigureAwait(false);
+            var referencedSymbols = progress.GetReferencedSymbols();
+            var referencedLocations = referencedSymbols.SelectMany(referencedSymbol => referencedSymbol.Locations).Distinct().ToImmutableArray()
+                .OrderByDescending(reference => reference.Location.SourceSpan.Start);
+
+            var list = new List<TInvocationExpressionSyntax>();
+            methodCallSites.Add(document.Document, list);
+            foreach (var refLocation in referencedLocations)
+            {
+                if (!methodCallSites.TryGetValue(refLocation.Document, out var invocations))
+                {
+                    invocations = new List<TInvocationExpressionSyntax>();
+                    methodCallSites.Add(refLocation.Document, invocations);
+                }
+
+                invocations.Add((TInvocationExpressionSyntax)refLocation.Location.FindNode(cancellationToken).GetRequiredParent());
+            }
+
+            return methodCallSites.ToImmutableDictionary();
+        }
+
+        private static SeparatedSyntaxList<SyntaxNode> AddArgumentToArgumentList(
+            SeparatedSyntaxList<SyntaxNode> invocationArguments, SyntaxGenerator generator,
+            SyntaxNode newArgumentExpression, int insertionIndex, string name, bool named)
+        {
+            SyntaxNode argument;
+            if (named)
+            {
+                argument = generator.Argument(name, RefKind.None, newArgumentExpression.WithAdditionalAnnotations(Simplifier.Annotation));
+            }
+            else
+            {
+                argument = generator.Argument(newArgumentExpression.WithAdditionalAnnotations(Simplifier.Annotation));
+            }
+
+            return invocationArguments.Insert(insertionIndex, argument);
+        }
+
+        private static ImmutableArray<SyntaxNode> AddExpressionArgumentToArgumentList(
+            ImmutableArray<SyntaxNode> arguments, SyntaxNode expression, SyntaxGenerator generator)
+        {
+            var newArgument = generator.Argument(expression);
+            return arguments.Add(newArgument);
+        }
+
+        /// <summary>
+        /// Generates a method declaration containing a return expression of the highlighted expression
+        /// </summary>
```

Comment:
i recommend a full example here.

Reply:
added

Reply:
i would just call this 'ExtractMethodXXX' (where XXX is optional, or whatever you think makes sense).

Reply:
changed